README.md
# Apache Tomcat Docker Images

Docker Image of Apache Tomcat maintained by XeniT.

Apache Tomcat is an open source web server and servlet container, by the Apache Software Foundation.

## Supported tags

* `alfresco-7.2-ubuntu`, `alfresco-7.1-ubuntu`, `alfresco-7.0-ubuntu`, `alfresco-6.2-ubuntu`, `alfresco-6.1-ubuntu`, `alfresco-6.0-ubuntu`, `8.5.59-jdk-11u9-bionic`, `8.5-jdk-11u9-bionic`, `8-jdk-11u9-bionic`, `8.5.59-bionic`, `8.5-bionic`, `8-bionic`
* `alfresco-5.2-ubuntu`, `alfresco-5.1-ubuntu`, `alfresco-5.0-ubuntu`, `7.0.106-jdk-8u272-bionic`, `7.0-jdk-8u272-bionic`, `7-jdk-8u272-bionic`, `7.0.106-bionic`, `7.0-bionic`, `7-bionic`
* `alfresco-4.2-ubuntu`, `7.0.106-jdk-7u211-trusty`, `7.0-jdk-7u211-trusty`, `7-jdk-7u211-trusty`

Additional tags are provided:
* tags containing a `-SNAPSHOT` postfix are development artifacts (from the master branch). Do not use them in a production environment.
* tags without any postfix are release artifacts

## Environment variables

There are several environment variables available to tweak the behaviour. While none of the variables are required, 
they may significantly aid you in using these images. The variables are read by an init script which further sets 
JAVA_OPTS variables to be used in tomcat's templated configuration file server.xml.

Environment variables:

| Variable                    |  Default                        | Java variable |
| --------------------------- | ------------------------------- | ---------------------------- |
| TOMCAT_PORT                 |  8080                           | -DTOMCAT_PORT                |
| TOMCAT_PORT_SSL             |  8443                           | -DTOMCAT_PORT_SSL            |
| TOMCAT_AJP_PORT             |  8009                           | -DTOMCAT_AJP_PORT            |
| TOMCAT_SERVER_PORT          |  8005                           | -DTOMCAT_SERVER_PORT         |
| TOMCAT_MAX_HTTP_HEADER_SIZE |  32768                          | -DTOMCAT_MAX_HTTP_HEADER_SIZE  or -DMAX_HTTP_HEADER_SIZE |
| TOMCAT_MAX_THREADS          |  200                            | -DTOMCAT_MAX_THREADS or -DMAX_THREADS |
| TOMCAT_ACCESSLOG_ROTATE     |  true                           | -DTOMCAT_ACCESSLOG_ROTATE    |
| TOMCAT_ACCESSLOG_MAXDAYS    |  -1                             | -DTOMCAT_ACCESSLOG_MAXDAYS   |
| TOMCAT_LOGGING_MAXDAYS      |  5                              | -DTOMCAT_LOGGING_MAXDAYS   |
| TOMCAT_RELAXED_QUERY_CHARS  |  ""                             | -DTOMCAT_RELAXED_QUERY_CHARS |
| TOMCAT_RELAXED_PATH_CHARS   |  ""                             | -DTOMCAT_RELAXED_PATH_CHARS  |


## Quick reference

**Where to get help:**

Channel [#docker](https://xenitengineering.slack.com/app_redirect?channel=docker) on Slack

**Maintained by:**

Toon Geens <toon.geens@xenit.eu>

**Supported Architectures:**

* linux-x64

**Supported Docker versions:**

* down to 1.10 (on a best-effort basis)

## Image variants

The `tomcat` image comes in a few flavors, where the variants use the following tag-structure:

```
tomcat:<version>[-<java>[-<os>]]
```

* **version**: the Tomcat version
* **java**: the Java distribution and version, for example `jdk-8` or `jdk-8u272-bionic`
* **os**: the Operating System, for example `bionic`

### Tomcat versions

In general, the latest update (minor version) of the last few major Tomcat releases are maintained:

* `9.0`
* `8.5` - _Tomcat 8.5_ is thought as a replacement for _Tomcat 8.0_. _Tomcat 8.0_ is [end of life](http://tomcat.apache.org/tomcat-80-eol.html) and is not supported.
* `7.0`

### Base images

All the available Tomcat images are built on top of the `docker.io/xenit/openjdk` images.

## Contributions

### How to build

Builds are produced by [GitHub Actions](https://github.com/xenit-eu/docker-tomcat/actions) driving Gradle from a
GitHub workflow file.

To build a local version of the _tomcat_ images:

```
./gradlew buildDockerImage
```



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)

## 2022-04

### Changed

* [2022-04-05] [PR #36](https://github.com/xenit-eu/docker-tomcat/pull/36) Update tomcat 8.5 & 9 to latest microversion to mitigate cve-2022-22965

## 2022-03

### Added

* [2022-03-30] [PR #32](https://github.com/xenit-eu/docker-tomcat/pull/31) Added alfresco-7.2-ubuntu tag

## 2022-02

### Added

* [2022-02-02] [PR #31](https://github.com/xenit-eu/docker-tomcat/pull/31) Added tomcat 9 build + alfresco-7.1-ubuntu tag

## 2021-04

### Changed

* [2021-12-10] [PR #29](https://github.com/xenit-eu/docker-tomcat/pull/29) Removed `centos` builds
* [2021-12-10] [PR #30](https://github.com/xenit-eu/docker-tomcat/pull/29) Use upstream images with unspecified jdk-minor version

## 2021-04

### Changed

* [2021-04-19] [PR #27](https://github.com/xenit-eu/docker-tomcat/pull/27) Support for application log deletion 	

## 2021-02

### Changed	
	
* [2021-02-22] [PR #23](https://github.com/xenit-eu/docker-tomcat/pull/23) Updated Java8 to jdk-8u275-b01
* [2021-02-01] [PR #20](https://github.com/xenit-eu/docker-tomcat/pull/20) Updated JDK to jdk-11.0.10

## 2020-12

### Changed

* [2020-12-05] [I-16](https://github.com/xenit-eu/docker-tomcat/issues/16) Move build from Travis to GitHub Actions

## 2020-10

### Changed

* [2020-10-28] [PR #14](https://github.com/xenit-eu/docker-tomcat/pull/14) Updated Tomcat to 7.0.106 / 8.5.59
* [2020-10-28] [PR #13](https://github.com/xenit-eu/docker-tomcat/pull/13) Updated JDK to jdk-11.0.9 and jdk-8u272 and for CentOS to jdk-7u261

## 2020-04

### Changed

* [2020-04-18] [PR #10](https://github.com/xenit-eu/docker-tomcat/pull/10) Updated JDK to [jdk-11.0.7](https://adoptopenjdk.net/release_notes.html?jvmVariant=hotspot#jdk11_0_7) and [jdk-8u252](https://adoptopenjdk.net/release_notes.html?jvmVariant=hotspot#jdk8u252)
* [2020-04-18] [PR #9](https://github.com/xenit-eu/docker-tomcat/pull/9) Updated Tomcat 8.5 to [8.5.54](http://tomcat.apache.org/tomcat-8.5-doc/changelog.html#Tomcat_8.5.54_(markt))

## 2020-03

### Changed

* [DOCKER-313] [PR #6](https://github.com/xenit-eu/docker-tomcat/pull/6) Updated Tomcat to 7.0.103 / 8.5.53 
* [DOCKER-297] [PR #4](https://github.com/xenit-eu/docker-tomcat/pull/4) Base JDK-8 ubuntu images on Bionic (18.04 LTS)
* [DOCKER-295] [PR #3](https://github.com/xenit-eu/docker-tomcat/pull/3) Make `relaxedQueryChars` and `relaxedPathChars` configurable
* [DOCKER-291] [PR #2](https://github.com/xenit-eu/docker-tomcat/pull/2) Updated Tomcat 8.5 to 8.5.49
* [DOCKER-123] [PR #1](https://github.com/xenit-eu/docker-tomcat/pull/1) Parametrize more variables in tomcat access valve: `maxDays` and `rotatable`


src/main/docker/85-init-tomcat.sh
#!/bin/bash

echo "Tomcat init start"

# the key is ignored, the value should contain the "-D" flag if it's a property
function setJavaOption {
    JAVA_OPTS="$JAVA_OPTS $2"
}

TOMCAT_CONTEXT_FILE=${CATALINA_HOME}'/conf/context.xml'
TOMCAT_CONFIG_FILE=${CATALINA_HOME}'/bin/setenv.sh'

# Tomcat-related properties
# for backwards compatibility, we keep the old options as well
setJavaOption 'TOMCAT_PORT' '-DTOMCAT_PORT='${TOMCAT_PORT:-8080}
setJavaOption 'TOMCAT_PORT_SSL' '-DTOMCAT_PORT_SSL='${TOMCAT_PORT_SSL:-8443}
setJavaOption 'TOMCAT_AJP_PORT' '-DTOMCAT_AJP_PORT='${TOMCAT_AJP_PORT:-8009}
setJavaOption 'TOMCAT_SERVER_PORT' '-DTOMCAT_SERVER_PORT='${TOMCAT_SERVER_PORT:-8005}
BC_TOMCAT_MAX_HTTP_HEADER_SIZE=${TOMCAT_MAX_HTTP_HEADER_SIZE:-$MAX_HTTP_HEADER_SIZE}
setJavaOption 'TOMCAT_MAX_HTTP_HEADER_SIZE' '-DTOMCAT_MAX_HTTP_HEADER_SIZE='${BC_TOMCAT_MAX_HTTP_HEADER_SIZE:-32768}
BC_TOMCAT_MAX_THREADS=${TOMCAT_MAX_THREADS:-$MAX_THREADS}
setJavaOption 'TOMCAT_MAX_THREADS' '-DTOMCAT_MAX_THREADS='${BC_TOMCAT_MAX_THREADS:-200}
setJavaOption 'TOMCAT_ACCESSLOG_ROTATE' '-DTOMCAT_ACCESSLOG_ROTATE='${TOMCAT_ACCESSLOG_ROTATE:-'true'}
setJavaOption 'TOMCAT_ACCESSLOG_MAXDAYS' '-DTOMCAT_ACCESSLOG_MAXDAYS='${TOMCAT_ACCESSLOG_MAXDAYS:-'-1'}
setJavaOption 'TOMCAT_RELAXED_QUERY_CHARS' '-DTOMCAT_RELAXED_QUERY_CHARS='${TOMCAT_RELAXED_QUERY_CHARS:-''}
setJavaOption 'TOMCAT_RELAXED_PATH_CHARS' '-DTOMCAT_RELAXED_PATH_CHARS='${TOMCAT_RELAXED_PATH_CHARS:-''}
setJavaOption 'TOMCAT_LOGGING_MAXDAYS' '-DTOMCAT_LOGGING_MAXDAYS='${TOMCAT_LOGGING_MAXDAYS:-'5'}

sed -i "s|\(<Context>\)|\1\n<Manager pathname=\"\" />|" ${TOMCAT_CONTEXT_FILE} || exit 1 # No persistent sessions after restart

echo "JAVA_OPTS=\"$JAVA_OPTS\"" >$TOMCAT_CONFIG_FILE
echo "export JAVA_OPTS" >> $TOMCAT_CONFIG_FILE

echo "Tomcat init done"
